controller/backup_controller.go

```diff
@@ -434,99 +510,139 @@ func (bc *BackupController) backupCreation(log logrus.FieldLogger, engineClient
-				backupStatusList := e.Status.BackupStatus
-				for _, b := range backupStatusList {
-					if b.SnapshotName == backup.Spec.SnapshotName {
-						bks = b
-						break
-					}
-				}
-			}
-			if bks == nil {
-				state = types.BackupStateUnknown
-				event(err, state, backup, volume)
-				return
-			}
-			if bks.Error != "" {
-				state = types.BackupStateError
-				event(errors.New(bks.Error), state, backup, volume)
-				return
-			}
+type BackupMonitor struct {
+	logger logrus.FieldLogger
 
-			if bks.Progress != 100 {
-				time.Sleep(BackupStatusQueryInterval)
-				continue
-			}
+	namespace          string
+	backupName         string
+	snapshotName       string
+	engineClient       engineapi.EngineClient
+	engineBackupStatus *types.EngineBackupStatus
 
-			// TODO:
-			//   use resource monitoring https://github.com/longhorn/longhorn/issues/2441
-			//   to trigger updates backup volume to run reconcile immediately
-			state = types.BackupStateCompleted
-			event(nil, state, backup, volume)
-
-			syncTime := time.Now().UTC()
-			backupVolume, err := bc.ds.GetBackupVolume(volumeName)
-			if err == nil {
-				// Request backup_volume_controller to reconcile BackupVolume immediately.
-				backupVolume.Spec.SyncRequestedAt = syncTime
-				if _, err = bc.ds.UpdateBackupVolume(backupVolume); err != nil && !apierrors.IsConflict(errors.Cause(err)) {
-					log.WithError(err).Errorf("Error updating backup volume %s spec", volume.Name)
-				}
-			} else if err != nil && apierrors.IsNotFound(err) {
-				// Request backup_target_controller to reconcile BackupTarget immediately.
-				backupTarget, err := bc.ds.GetBackupTarget(types.DefaultBackupTargetName)
-				if err != nil {
-					log.WithError(err).Warn("Failed to get backup target")
-					return
-				}
-				backupTarget.Spec.SyncRequestedAt = syncTime
-				if _, err = bc.ds.UpdateBackupTarget(backupTarget); err != nil && !apierrors.IsConflict(errors.Cause(err)) {
-					log.WithError(err).Warn("Failed to update backup target")
+	lock  sync.Mutex
+	stale bool
+
+	syncCallback func(key string)
+
+	ctx  context.Context
+	quit context.CancelFunc
+}
+
+func NewBackupMonitor(logger logrus.FieldLogger, backup *longhorn.Backup, engineClient engineapi.EngineClient, syncCallback func(key string)) *BackupMonitor {
+	ctx, quit := context.WithCancel(context.Background())
+	monitor := &BackupMonitor{
+		logger: logger,
+
+		namespace:          backup.Namespace,
+		backupName:         backup.Name,
+		snapshotName:       backup.Spec.SnapshotName,
+		engineClient:       engineClient,
+		engineBackupStatus: &types.EngineBackupStatus{},
+
+		lock:  sync.Mutex{},
+		stale: true,
+
+		syncCallback: syncCallback,
+
+		ctx:  ctx,
+		quit: quit,
+	}
+
+	go monitor.monitorBackups()
+
+	return monitor
+}
+
+func (m *BackupMonitor) Close() {
+	m.quit()
+}
+
+func (m *BackupMonitor) syncBackups() error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	engineBackupStatusList, err := m.engineClient.SnapshotBackupStatus()
```

Comment:
Just notice that here the output of this function is a map. But the result from the cmd is a struct. Can you double-check this part?
https://github.com/longhorn/longhorn-engine/blob/31ac10b371897346f9f3b08a731af48f5f132d70/app/cmd/backup.go#L229

Reply:
The result of the cmd is a raw string, for example:
```shell
{
	"backup-02b651e23f314cf3": {
		"progress": 100,
		"backupURL": "s3://backupbucket@minio/?backup=backup-02b651e23f314cf3\u0026volume=pvc-75631cf4-cd4e-40e7-b05f-acd87efdda0d",
		"snapshotName": "78684fb5-696f-4071-a1f6-4a5d44444d47",
		"state": "complete",
		"replicaAddress": "tcp://10.42.2.115:10000"
	},
	"backup-a2471feb35b04a1d": {
		"progress": 100,
		"backupURL": "s3://backupbucket@minio/?backup=backup-a2471feb35b04a1d\u0026volume=pvc-75631cf4-cd4e-40e7-b05f-acd87efdda0d",
		"snapshotName": "ebc45a24-c226-4c2f-b55e-732f8e004b7b",
		"state": "complete",
		"replicaAddress": "tcp://10.42.2.115:10000"
	}
}
```

Longhorn manager [unmarshal](https://github.com/longhorn/longhorn-manager/blob/f83a22c/engineapi/backups.go#L300-L301) the console output to the map.